Week 06/Lecture 11/Assignment 01/employee/src/main/java/com/example/employee/service/impl/EmployeeServiceImpl.java

```diff
@@ -0,0 +1,177 @@
+     * @return page of employee DTO
+     */
+    @Override
+    @Transactional
+    public Page<EmployeeDTO> getAllEmployees(Pageable pageable) {
+        Page<Employee> employees = employeeRepository.findAll(pageable);
+        return employees.map(this::mapToEmployeeDTO);
+    }
+
+    /**
+     * Get employee by id
+     * @param id employee id
+     * @return employee DTO
+     */
+    @Override
+    public EmployeeDTO getEmployeeById(Integer id) {
+        Employee employee = employeeRepository.findById(id).orElseThrow();
+        return mapToEmployeeDTO(employee);
+    }
+
+    /**
+     * Add employee
+     * @param employeeDTO employee information
+     * @return employee DTO
+     */
+    @Override
+    public EmployeeDTO addEmployee(EmployeeDTO employeeDTO) {
+        Employee employee = mapToEmployeeEntity(employeeDTO);
+        employee = employeeRepository.save(employee);
+        return mapToEmployeeDTO(employee);
+    }
+
+    /**
+     * Update employee
+     * @param id employee id
+     * @param employeeDTO employee information
+     * @return updated employee DTO
+     */
+    @Override
+    public EmployeeDTO updateEmployee(Integer id, EmployeeDTO employeeDTO) {
+        Employee employee = employeeRepository.findById(id).orElseThrow();
+        employee.setBirthDate(employeeDTO.getBirthDate());
+        employee.setFirstName(employeeDTO.getFirstName());
+        employee.setLastName(employeeDTO.getLastName());
+        employee.setGender(employeeDTO.getGender());
+        employee.setHireDate(employeeDTO.getHireDate());
+        employeeRepository.save(employee);
+        return mapToEmployeeDTO(employee);
+    }
+
+    /**
+     * Delete employee
+     * @param id employee id
+     */
+    @Override
+    public void deleteEmployee(Integer id) {
+        employeeRepository.deleteById(id);
+    }
+
+    /**
+     * Search employees
+     * @param criteria search criteria
+     * @param pageable pagination information
+     * @return page of employee DTO
+     */
+    @Override
+    public Page<EmployeeDTO> searchEmployees(EmployeeSearchCriteria criteria, Pageable pageable) {
+        EmployeeSpecification specification = new EmployeeSpecification(criteria);
+        Page<Employee> employees = employeeRepository.findAll(specification, pageable);
+        return employees.map(this::mapToEmployeeDTO);
+    }
+
+    /**
+     * Map employee entity to employee DTO
+     * @param employee employee entity
+     * @return employee DTO
+     */
+    private EmployeeDTO mapToEmployeeDTO(Employee employee) {
+        EmployeeDTO employeeDTO = new EmployeeDTO();
+        employeeDTO.setEmpNo(employee.getEmpNo());
+        employeeDTO.setBirthDate(employee.getBirthDate());
+        employeeDTO.setFirstName(employee.getFirstName());
+        employeeDTO.setLastName(employee.getLastName());
+        employeeDTO.setGender(employee.getGender());
+        employeeDTO.setHireDate(employee.getHireDate());
```

Comment:
Can use lombok

Reply:
Thankyou for the feedback, I've added the Lombok's `@Builder` and `@Data` annotations for mapping method of DTO -> entity and vice versa.